packages/aws-cdk-lib/aws-iam/lib/principals.ts

```diff
@@ -473,14 +473,16 @@ export class ArnPrincipal extends PrincipalBase {
 export class AccountPrincipal extends ArnPrincipal {
   public readonly principalAccount: string | undefined;
 
+  private accountIdRegExp = new RegExp('^[0-9]{12}$');
+
   /**
    *
    * @param accountId AWS account ID (i.e. '123456789012')
    */
   constructor(public readonly accountId: any) {
     super(new StackDependentToken(stack => `arn:${stack.partition}:iam::${accountId}:root`).toString());
-    if (!cdk.Token.isUnresolved(accountId) && typeof accountId !== 'string') {
-      throw new Error('accountId should be of type string');
+    if (!cdk.Token.isUnresolved(accountId) && typeof accountId !== 'string' && !this.accountIdRegExp.test(accountId)) {
```

Comment:
While you're at it, maybe we should check to see that all characters in the string are digits.

Reply:
It's covered with the regex `new RegExp('^[0-9]{12}$');`